Dashboard save button

graceguo-supercat opened the pull request:
![tr6goydv67](https://user-images.githubusercontent.com/27990562/39950499-50f79cbe-5536-11e8-809e-c0986894f524.gif)


- save dashboard button
- bring back save custom css feature
- also hide/disable button when user didn't have access:

- for read-only user:
<img width="1447" alt="screen shot 2018-05-22 at 3 48 38 pm" src="https://user-images.githubusercontent.com/27990562/40394328-d6d3a748-5dd7-11e8-8a1f-7c1af6151e3b.png">

- for user can copy dashboard (not the owner of dashboard): they can save a new copy then edit on his new copy. `Edit Dashboard` and `Overwrite dashboard` options are disabled:
<img width="1267" alt="screen shot 2018-05-22 at 3 57 38 pm" src="https://user-images.githubusercontent.com/27990562/40394686-4ff6c596-5dd9-11e8-8498-9d7daac296a5.png">

<img width="1261" alt="screen shot 2018-05-22 at 3 59 00 pm" src="https://user-images.githubusercontent.com/27990562/40394628-1339d6de-5dd9-11e8-9130-669ad822162d.png">


- for user has full access to dashboard: they can edit, save, save-as:
<img width="1266" alt="screen shot 2018-05-22 at 4 02 29 pm" src="https://user-images.githubusercontent.com/27990562/40394750-95fdc74c-5dd9-11e8-92ec-1c088b3c42c6.png">

williaster commented on superset/assets/src/dashboard/actions/dashboardState.js:
```diff
@@ -81,14 +81,28 @@ export function onSave() {
+      data: {
+        data: JSON.stringify(data),
+      },
+      success: () => dispatch(saveDashboardRequestSuccess()),
```
what about an error?

graceguo-supercat replied:
added error handle.

williaster commented on superset/assets/src/dashboard/components/Header.jsx:
```diff
@@ -88,6 +100,36 @@ class Header extends React.PureComponent {
   }
 
+  overwriteDashboard() {
+    const dashboardId = this.props.dashboardInfo.id;
```
why not pull this out with the rest of the `props` destructuring you're doing below, and then just use `dashboardInfo.id` in the `onSave` call? (just strange to do partial destructuring)

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/components/Header.jsx:
```diff
@@ -88,6 +100,36 @@ class Header extends React.PureComponent {
+      default_filters: JSON.stringify(filters),
+    };
+
+    this.props
```
not a blocker, but it would be better to move all of this logic to a separate action instead of overloading a component.

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/components/SaveModal.jsx:
```diff
@@ -16,6 +15,7 @@ const propTypes = {
   dashboardTitle: PropTypes.string.isRequired,
   expandedSlices: PropTypes.object.isRequired,
   layout: PropTypes.object.isRequired,
+  saveType: PropTypes.string,
```
if you add a string prop you should _always_ enumerate all valid options.

graceguo-supercat replied:
add enumeration.

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -38,6 +38,22 @@
 }
 
+.dashboard .dashboard-header {
+  #save-dash-split-button {
```
does this have to be an id?

graceguo-supercat replied:
it's react Dropdown required.

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -38,6 +38,22 @@
+  #save-dash-split-button {
+    border-radius: 0;
+    margin-left: -5px;
+    padding: 6px;
```
I'm surprised this is 6px, most everything should be a multiple of 8px

graceguo-supercat replied:
this padding is not necessary. i removed it.

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -38,6 +38,22 @@
+    width: 30px;
+
+    .caret {
+      position: absolute;
```
can you avoid absolute positioning? like use flexbox?

graceguo-supercat replied:
.caret inherit many styles, if not use `position: absolute;` it's just too many things to overwrite.

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -38,6 +38,22 @@
+
+    .caret {
+      position: absolute;
+      top: 24px;
```
it'd be better to use flexbox or percentages so this doesn't break when the container changes size?

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -63,7 +79,7 @@
   position: absolute;
   top: 0;
-  right: -22px;
+  right: -16px;
```
can you add a comment that you're doing this to increase the clickable area for the button?

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -80,12 +96,17 @@
+  a[role="menuitem"] & {
+    width: 8px;
+    height: 8px;
+    margin-right: 6px;
```
what's the `margin-right` for?

graceguo-supercat replied:
this is the margin for the big red dot indicating it's cached data. it should be 8px. fixed.

williaster commented on superset/assets/src/dashboard/components/Header.jsx:
```diff
@@ -1,12 +1,18 @@
 import Button from '../../components/Button';
 import FaveStar from '../../components/FaveStar';
-import SaveModal from './SaveModal';
+import SaveModal, { SAVE_TYPE_OVERWRITE } from './SaveModal';
```
I think `SAVE_TYPE_OVERWRITE ` should be moved into the dashboard constants file, this defeats the purpose of having that file and centralizes all constants.

williaster commented on superset/assets/src/dashboard/components/SaveModal.jsx:
```diff
@@ -1,36 +1,40 @@
 import { t } from '../../locales';
 import Checkbox from '../../components/Checkbox';
 
+export const SAVE_TYPE_OVERWRITE = 'overwrite';
```
see comment about moving these to constants file.

williaster commented on superset/assets/src/dashboard/components/Header.jsx:
```diff
@@ -161,44 +197,63 @@ class Header extends React.PureComponent {
-                  dashboardId={this.props.dashboardInfo.id}
+                  dashboardId={dashboardInfo.id}
                   dashboardTitle={dashboardTitle}
+                  saveType="newDashboard"
```
should this also be one of the constants?

williaster commented on superset/assets/src/dashboard/components/SaveModal.jsx:
```diff
@@ -59,37 +64,7 @@ class SaveModal extends React.PureComponent {
   handleNameChange(event) {
     this.setState({
       newDashName: event.target.value,
-      saveType: 'newDashboard',
```
thanks for moving this to the new file!

williaster commented on superset/assets/src/dashboard/components/SaveModal.jsx:
```diff
@@ -138,15 +112,15 @@ class SaveModal extends React.PureComponent {
-              value="overwrite"
+              value={SAVE_TYPE_OVERWRITE}
               onChange={this.handleSaveTypeChange}
               checked={this.state.saveType === 'overwrite'}
```
should this be using the constant?

williaster commented on superset/assets/src/dashboard/components/SaveModal.jsx:
```diff
@@ -138,15 +112,15 @@ class SaveModal extends React.PureComponent {
-              value="newDashboard"
+              value={SAVE_TYPE_NEWDASHBOARD}
               onChange={this.handleSaveTypeChange}
               checked={this.state.saveType === 'newDashboard'}
```
should this be using the constant?

williaster commented on superset/assets/src/dashboard/reducers/dashboardState.js:
```diff
@@ -14,13 +14,13 @@ import {
 
 export default function dashboardStateReducer(state = {}, action) {
   const actionHandlers = {
-    [UPDATE_DASHBOARD_TITLE]() {
```
nice catch

williaster reviewed (commented):
@graceguo-supercat this looks great overall, a couple more minor comments about using the constants and then I think it's ready to go.

Commit:
save button

Commit:
fix slices list height

Commit:
save custom css

Commit:
merge save-dash changes from dashboard v1
https://github.com/apache/incubator-superset/pull/4900
https://github.com/apache/incubator-superset/pull/5051

williaster commented:
LGTM! thanks for all the effort on this one and addressing the new v1 stuff 🔥 👍
